infrastructure/src/main/java/org/corfudb/infrastructure/orchestrator/actions/RestoreRedundancyMergeSegments.java

```diff
@@ -1,38 +1,219 @@
+     *
+     * @param stateList A list of transfer segments after a state transfer is done.
+     * @param runtime   A corfu runtime.
+     * @return A status RESTORED, if the transferred segments of the state list are now present
+     * in the accepted layout, NOT_RESTORED otherwise.
+     */
+    RestoreStatus restoreWithBackOff(List<TransferSegment> stateList,
+                                     CorfuRuntime runtime) throws Exception {
+
+        Consumer<ExponentialBackoffRetry> retrySettings = settings -> {
+            settings.setBase(retryBase);
+            settings.setExtraWait(extraWaitMillis);
+            settings.setBackoffDuration(Duration.ofSeconds(
+                    backoffDurationSeconds));
+            settings.setRandomPortion(randomPart);
+        };
+
+        AtomicInteger retries = new AtomicInteger(restoreRetries);
+        return IRetry.build(ExponentialBackoffRetry.class, RetryExhaustedException.class, () -> {
+            try {
+
+                runtime.invalidateLayout();
+                Layout oldLayout = runtime.getLayoutView().getLayout();
+                LayoutManagementView layoutManagementView = runtime.getLayoutManagementView();
+                return tryRestoreRedundancyAndMergeSegments(stateList, oldLayout, layoutManagementView);
+            } catch (WrongEpochException | QuorumUnreachableException | OutrankedException e) {
+                log.warn("Got: {}. Retrying: {} times.", e.getMessage(), retries.get());
+                if (retries.decrementAndGet() < 0) {
+                    log.error("Retries exhausted.");
+                    throw new RetryExhaustedException("Retries exhausted.");
+                } else {
+                    throw new RetryNeededException();
+                }
+            }
+        }).setOptions(retrySettings).run();
+
+    }
+
+    /**
+     * Try restore redundancy for all the currently transferred segments on the current node.
+     *
+     * @param stateList            A list that holds the state for every segment.
+     * @param oldLayout            A latest layout.
+     * @param layoutManagementView A layout view.
+     * @return A status of restoration: failed or succeeded.
+     */
+    RestoreStatus tryRestoreRedundancyAndMergeSegments(
+            List<TransferSegment> stateList, Layout oldLayout,
+            LayoutManagementView layoutManagementView) {
+
+        // Get all the transfers that failed.
+        List<TransferSegment> failed = stateList.stream()
+                .filter(segment -> segment.getStatus().getSegmentState() == FAILED)
+                .collect(Collectors.toList());
+
+        // Throw the first failure if present.
+        Optional<TransferSegmentException> transferSegmentFailure = failed.stream()
+                .findFirst()
+                .flatMap(ts -> ts.getStatus().getCauseOfFailure());
+
+        transferSegmentFailure.ifPresent(failure -> {
```

Comment:
why not using matchAny to test whether there is any failure?

Reply:
I only need a single failure in the list to throw an exception + findFirst is composed well with a getCauseOfFailure.